src/rust/ensogl/src/app/command.rs

```diff
@@ -0,0 +1,280 @@
+            let caption = command_doc_map.get(&t.label).unwrap().clone(); // fixme unwrap
+            let frp     = t.frp;
+            let endpoint = FrpEndpoint {caption,frp};
+            (t.label,endpoint)
+        }).collect();
+
+        let status_doc_map : HashMap<String,String> = T::status_api_docs().into_iter().map(|t| {
+            (t.label,t.caption)
+        }).collect();
+        let status_map = T::status_api(target).into_iter().map(|t| {
+            let caption = status_doc_map.get(&t.label).unwrap().clone(); // fixme unwrap
+            let frp     = t.frp;
+            let endpoint = FrpEndpoint {caption,frp};
+            (t.label,endpoint)
+        }).collect();
+
+        let instance = ProviderInstance {network,command_map,status_map};
+        let was_registered = self.instances.borrow().get(label).is_some();
+        if !was_registered {
+            self.register::<T>();
+            warning!(&self.logger,
+                "The command provider '{label}' was created but never registered. You should \
+                always register available command providers as soon as possible to spread the \
+                information about their API.");
+        };
+        self.instances.borrow_mut().get_mut(label).unwrap().push(instance);
+    }
+}
+
+
+
+// ==============
+// === Macros ===
+// ==============
+
+/// Defines new `Status` API. See docs of `StatusApi` to learn more.
+#[macro_export]
+macro_rules! def_status_api {
+    ( $name:ident
+        $(
+            #[doc=$($doc:tt)*]
+            $field:ident
+        ),* $(,)?
+    ) => {
+        #[derive(Debug,Clone,CloneRef)]
+        pub struct $name {
+            $(pub $field : frp::Sampler<bool>),*
+        }
+
+        impl app::command::StatusApi for $name {
+            fn status_api_docs() -> Vec<app::command::EndpointDocs> {
+                vec! [$(app::command::EndpointDocs::new(stringify!($field),$($doc)*)),*]
+            }
+
+            fn status_api(&self) -> Vec<app::command::StatusEndpoint> {
+                vec! [$(app::command::StatusEndpoint::new(stringify!($field),&self.$field)),*]
+            }
+        }
+    };
+}
+
+/// Defines new `Command` API. See docs of `CommandApi` to learn more.
+#[macro_export]
+macro_rules! def_command_api {
+    ( $name:ident
+        $(
+            #[doc=$($doc:tt)*]
+            $field:ident
+        ),* $(,)?
+    ) => {
+        #[derive(Debug,Clone,CloneRef)]
+        pub struct $name {
+            $(pub $field : frp::Source),*
+        }
+
+        impl app::command::CommandApi for $name {
+            fn command_api_docs() -> Vec<app::command::EndpointDocs> {
+                vec! [$(app::command::EndpointDocs::new(stringify!($field),$($doc)*)),*]
+            }
+
+            fn command_api(&self) -> Vec<app::command::CommandEndpoint> {
+                vec! [$(app::command::CommandEndpoint::new(stringify!($field),&self.$field)),*]
+            }
+        }
+    };
+}
```

Comment:
I don't see _any_ tests for registering modules which seems to be a huge logic.

Reply:
As agreed via phone - tests will be added in next prs.